File: cpp/cpp_offer/offerII091.cc
// C++
// leetcode offerII 091
// https://leetcode.cn/problems/JEj789/

class Solution {
public:
    int minCost(vector<vector<int>>& costs) {
    	int n = costs.size(), res = INT_MAX;
    	for (int i = 1; i < n; i++) {
    		for(int j = 0; j < 3; j++) {
    			int cur = INT_MAX;
    			for(int k = 0; k < 3; k++) {
    				if (j == k)
    					continue;
    				cur = min(cur, costs[i][j] + costs[i-1][k]);
    			}
    			costs[i][j] = cur;
    		}
    	}
    	for(int i = 0; i < 3; i++) {
	    	res = min(res, costs[n-1][i]);
	    }
	    return res;
    }	
};

File: Go/Go_offer/offerII091.go
// Go
// leetcode offerII 091
// https://leetcode.cn/problems/JEj789/

func minCost(cs [][]int) int {
	n, res := len(cs), 30000
	
	// dp
	for i := 1; i < n; i++ {
		for j := 0; j < 3; j++ {
			cur := 30000
			for k := 0; k < 3; k++ {
				if j == k {
					continue
				}
				if now := cs[i][j] + cs[i-1][k]; now < cur {
					cur = now
				}
			}
			cs[i][j] = cur
		}
	}

	// collecting answer from row n-1 
	for i := 0; i < 3; i++ {
		if cs[n-1][i] < res {
			res = cs[n-1][i]
		}
	}
	return res
}